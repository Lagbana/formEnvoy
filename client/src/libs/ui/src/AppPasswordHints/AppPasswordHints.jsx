/* Password hints
=============================== */

import React from 'react';
import { StyleSheet, Text, View, ViewStyle } from 'react-native';
import { Typography, Colors } from 'libs/ui/src/WebButton/node_modules/libs/ui/src/WebLoading/node_modules/colibri/styles';
import { AppIcon } from '../AppIcon';
import { useTranslations } from '../useTranslations';

export interface AppPasswordHintsRule {
	label: string;
	vars: { num: number };
	check: (t: string) => boolean;
}

export interface AppPasswordHintsProps {
	rules?: AppPasswordHintsRule[];
	password: string;
	visible?: boolean;
	invalidColor?: string;
	validColor?: string;
	style?: ViewStyle;
}

export const AppPasswordHints: React.FC<AppPasswordHintsProps> = ({
  password,
  visible = false,
  invalidColor = 'gray',
	validColor = 'green',
	rules = [],
  ...props
}) => {
  const { translate } = useTranslations();

  // styles

  const styles = StyleSheet.create({
    container: {
      flexDirection: 'row',
      flexWrap: 'wrap',
      marginTop: 8,
    },
    rule: {
      width: '50%',
      flexDirection: 'row',
      alignItems: 'center',
      paddingVertical: 2,
    },
    validRule: {},
    icon: {
      width: 16,
      height: 16,
      marginRight: 8,
      color: Colors[invalidColor],
      opacity: 0.5,
    },
    validIcon: {
      color: Colors[validColor],
      opacity: 1,
    },
    label: {
      ...Typography.labelXsmall,
      color: Colors[invalidColor],
    },
    validLabel: {
      color: Colors[validColor],
    },
  });

  return (
    <>
      {visible && (
        <View {...props} style={[styles.container, props.style]}>
          {rules.map(({ label, vars, check }, index) => (
            <View
              key={index}
              style={[styles.rule, check(password) ? styles.validRule : '']}
            >
              <AppIcon
                name={check(password) ? 'check' : 'x'}
                size={16}
                style={[
                  styles.icon,
                  check(password) ? styles.validIcon : undefined,
                ]}
              />
              <Text
                style={[
                  styles.label,
                  check(password) ? styles.validLabel : undefined,
                ]}
              >
                {translate(label, vars)}
              </Text>
            </View>
          ))}
        </View>
      )}
    </>
  );
};
