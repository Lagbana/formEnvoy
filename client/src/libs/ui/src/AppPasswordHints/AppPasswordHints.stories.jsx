import React from 'react'
import { storiesOf } from '@storybook/react-native'
import { AppPasswordHints } from './AppPasswordHints'

storiesOf('AppPasswordHints', module)
  .add('Visible: True', () => (
    <AppPasswordHints
      visible={true}
      password='abc123'
      children={'Hello Password Hints!'}
      invalidColor={'blue'}
      validColor={'red'}
    />
  ))

  .add('Visible: False', () => (
    <AppPasswordHints
      visible={false}
      password='abc123'
      children={'Hello Password Hints!'}
      invalidColor={'blue'}
      validColor={'red'}
    />
  ))

  .add('With rules', () => (
    <AppPasswordHints
      rules={[
        {
          label: 'password',
          vars: { num: 8 },
          check: (t: 'checked') => {
            return true
          }
        }
      ]}
      visible={false}
      password='abc123'
      children={'Hello Password Hints!'}
      invalidColor={'blue'}
      validColor={'red'}
    />
  ))
