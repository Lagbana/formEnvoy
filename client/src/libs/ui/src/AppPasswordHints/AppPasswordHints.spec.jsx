import React from 'react'
import { shallow, configure } from 'enzyme'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'
import { AppPasswordHints } from './AppPasswordHints'

configure({ adapter: new Adapter() })

test('<AppPasswordHints /> renders correctly', () => {
  const component = shallow(<AppPasswordHints password={'password123'} />)
  component.unmount()
  expect(toJson(component)).toMatchSnapshot()
})
