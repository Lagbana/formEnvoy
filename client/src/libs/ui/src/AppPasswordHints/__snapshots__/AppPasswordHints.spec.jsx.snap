// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppPasswordHints /> renders correctly 1`] = `""`;
